tests/src/pcm_loopback_test.cc

```diff
@@ -0,0 +1,219 @@
+
+        channels[0] = 0.0;
+        for (int32_t i = 1; i < CH; ++i) {
+            channels[i] = channels[i - 1] + phase;
+        }
+
+        gain = std::pow(M_E, std::log(10) * static_cast<double>(L) / 20.0);
+    }
+
+    ~SineToneGenerator() = default;
+
+    int32_t Read(void *buffer, int32_t size) override {
+        Type *pcm_buffer = reinterpret_cast<Type *>(buffer);
+
+        size = (size / (CH * sizeof(Type))) * (CH * sizeof(Type));
+        int32_t samples = size / sizeof(Type);
+        int32_t s = 0;
+
+        while (s < samples) {
+            for (int32_t i = 0; i < CH; ++i) {
+                pcm_buffer[s++] = GetSample(channels[i]);
+                channels[i] += kStep;
+            }
+        }
+        return size;
+    }
+};
+
+template<typename T>
+static double Energy(T *buffer, size_t samples) {
+    double sum = 0.0;
+    for (size_t i = 0; i < samples; i++) {
+        sum += static_cast<double>(buffer[i]) * static_cast<double>(buffer[i]);
+    }
+    return sum;
+}
+
+TEST(PcmLoopbackTest, LoopbackS16le) {
+    static constexpr unsigned int kDefaultChannels = 2;
+    static constexpr unsigned int kDefaultSamplingRate = 48000;
+    static constexpr unsigned int kDefaultPeriodSize = 1024;
+    static constexpr unsigned int kDefaultPeriodCount = 3;
+    static constexpr pcm_config kInConfig = {
+        .channels = kDefaultChannels,
+        .rate = kDefaultSamplingRate,
+        .period_size = kDefaultPeriodSize,
+        .period_count = kDefaultPeriodCount,
+        .format = PCM_FORMAT_S16_LE,
+        .start_threshold = 0,
+        .stop_threshold = 0,
+        .silence_threshold = 0,
+        .silence_size = 0,
+    };
+    pcm *pcm_in = pcm_open(kLoopbackCard, kLoopbackCaptureDevice, PCM_IN, &kInConfig);
+    ASSERT_TRUE(pcm_is_ready(pcm_in));
+
+    static constexpr pcm_config kOutConfig = {
+        .channels = kDefaultChannels,
+        .rate = kDefaultSamplingRate,
+        .period_size = kDefaultPeriodSize,
+        .period_count = kDefaultPeriodCount,
+        .format = PCM_FORMAT_S16_LE,
+        .start_threshold = kDefaultPeriodSize,
+        .stop_threshold = kDefaultPeriodSize * kDefaultPeriodCount,
+        .silence_threshold = 0,
+        .silence_size = 0,
+    };
+    pcm *pcm_out = pcm_open(kLoopbackCard, kLoopbackPlaybackDevice, PCM_OUT, &kOutConfig);
+    ASSERT_TRUE(pcm_is_ready(pcm_out));
+
+    ASSERT_EQ(pcm_link(pcm_in, pcm_out), 0);
+
+    bool stopping = false;
+    ASSERT_EQ(pcm_get_subdevice(pcm_in), pcm_get_subdevice(pcm_out));
+
+    std::thread capture([pcm_in, &stopping] {
+        size_t buffer_size = pcm_frames_to_bytes(pcm_in, kDefaultPeriodSize);
+        unsigned int frames = pcm_bytes_to_frames(pcm_in, buffer_size);
+        auto buffer = std::make_unique<unsigned char[]>(buffer_size);
+        int32_t counter = 0;
+        while (!stopping) {
+            EXPECT_EQ(pcm_readi(pcm_in, buffer.get(), frames), frames);
+            counter++;
+            if (counter >= 5) {
```

Comment:
Can you please add comment to explain why it is 5 ?

Reply:
Done.